File: example-inference-llm.mjs
#!/usr/bin/env node
/**
 * example-inference-llm.mjs — End-to-end GGUF LLM inference via model-resolver-node
 *
 * Demonstrates:
 *   1. Resolving a GGUF model from a local flat-repo (or CDN)
 *   2. Loading with node-llama-cpp (npm install node-llama-cpp)
 *   3. Single-turn and multi-turn chat with LlamaChatSession
 *
 * Usage:
 *   node example-inference-llm.mjs --source /path/to/pkg-gemma3
 *
 *   # CDN source:
 *   node example-inference-llm.mjs \
 *     --source https://cdn.jsdelivr.net/gh/user/cdn-llm@v1 \
 *     --manifest q4_0
 *
 * Requirements:
 *   npm install node-llama-cpp
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// ─── Parse args ────────────────────────────────────────────────────────────

const argv = process.argv.slice(2);
const getArg = (name, fallback) => {
  const i = argv.indexOf(name);
  return i >= 0 && argv[i + 1] ? argv[i + 1] : fallback;
};

const SOURCE   = getArg('--source', null);
const MANIFEST = getArg('--manifest', 'q4_0');
const CACHE_DIR = getArg('--cache-dir', './.model-cache');
const MAX_TOKENS = parseInt(getArg('--max-tokens', '128'), 10);

if (!SOURCE) {
  console.log('Usage: node example-inference-llm.mjs --source /path/to/pkg-gemma3');
  process.exit(1);
}

// ─── Main ──────────────────────────────────────────────────────────────────

async function main() {
  const { ModelResolver } = await import(path.join(__dirname, 'model-resolver-node.mjs'));
  const { getLlama, LlamaChatSession } = await import('node-llama-cpp');

  const resolver = new ModelResolver({ cacheDir: CACHE_DIR });

  console.log('='.repeat(60));
  console.log('LLM INFERENCE via model-resolver-node + node-llama-cpp');
  console.log('='.repeat(60));

  // ── 1. Resolve GGUF model ─────────────────────────────────────────────

  console.log(`\nResolving GGUF model from: ${SOURCE}`);
  let t0 = Date.now();
  const files = await resolver.resolveFiles(SOURCE, {
    manifest: MANIFEST,
    onProgress: (p) => process.stdout.write(
      `\r  [${String(p.percent).padStart(3)}%] ${(p.file || '').slice(0, 50).padEnd(50)}`),
  });

  const ggufPath = Object.entries(files)
    .filter(([vp]) => vp.endsWith('.gguf'))
    .map(([, fp]) => fp)[0];

  if (!ggufPath) { console.error('No .gguf file found in manifest'); process.exit(1); }

  const sizeMb = (fs.statSync(ggufPath).size / 1048576).toFixed(1);
  console.log(`\n  → ${ggufPath}  (${((Date.now() - t0) / 1000).toFixed(1)}s)`);
  console.log(`  Size: ${sizeMb} MB`);

  // ── 2. Load model ─────────────────────────────────────────────────────

  console.log('\nLoading model with node-llama-cpp...');
  t0 = Date.now();
  const llama = await getLlama({ gpu: false });
  const model = await llama.loadModel({ modelPath: ggufPath });
  console.log(`  ✓ Model loaded (${((Date.now() - t0) / 1000).toFixed(1)}s)`);

  // ── 3. Single-turn inference ──────────────────────────────────────────

  console.log('\n' + '='.repeat(60));
  console.log('SINGLE-TURN INFERENCE');
  console.log('-'.repeat(60));

  const prompts = [
    'What is 2 + 2? Answer in one sentence.',
    'Explain gravity to a 5 year old in two sentences.',
    'Write a haiku about programming.',
  ];

  for (const prompt of prompts) {
    // Fresh context+session per prompt for single-turn
    const sCtx = await model.createContext({ contextSize: 512 });
    const session = new LlamaChatSession({ contextSequence: sCtx.getSequence() });
    console.log(`\n  User: ${prompt}`);
    t0 = Date.now();
    const response = await session.prompt(prompt, { maxTokens: MAX_TOKENS, temperature: 0.7 });
    console.log(`  Model: ${response.trim()}`);
    console.log(`  (${((Date.now() - t0) / 1000).toFixed(1)}s)`);
    await sCtx.dispose();
  }

  // ── 4. Multi-turn conversation ────────────────────────────────────────

  console.log('\n' + '='.repeat(60));
  console.log('MULTI-TURN CONVERSATION');
  console.log('-'.repeat(60));

  // node-llama-cpp LlamaChatSession maintains history automatically
  const chatCtx = await model.createContext({ contextSize: 2048 });
  const chat = new LlamaChatSession({ contextSequence: chatCtx.getSequence() });

  const turns = [
    'What is the capital of France?',
    'What is its population?',
    'And what is a famous landmark there?',
  ];

  for (const turn of turns) {
    console.log(`\n  User: ${turn}`);
    const reply = await chat.prompt(turn, { maxTokens: 64, temperature: 0.3 });
    console.log(`  Model: ${reply.trim()}`);
  }

  // ── Cleanup ───────────────────────────────────────────────────────────

  await chatCtx.dispose();
  await model.dispose();
  await llama.dispose();

  console.log('\n' + '='.repeat(60));
  console.log('✓ All LLM inference complete');
  console.log('='.repeat(60));
}

main().catch(e => { console.error('FATAL:', e.message); process.exit(1); });

File: example-inference-llm.py
#!/usr/bin/env python3
"""
example-inference-llm.py — End-to-end GGUF LLM inference via model_resolver

Demonstrates:
  1. Resolving a GGUF model from a local flat-repo (or CDN) using model_resolver
  2. Loading with llama-cpp-python (pip install llama-cpp-python)
  3. Single-turn and multi-turn chat completion

Usage:
  python example-inference-llm.py --source /path/to/pkg-gemma3

  # CDN source:
  python example-inference-llm.py \
    --source https://cdn.jsdelivr.net/gh/user/cdn-llm@v1 \
    --manifest q4_0

Requirements:
  pip install llama-cpp-python
"""

import argparse, os, sys, time

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from model_resolver import ModelResolver, resolve_gguf

try:
    from llama_cpp import Llama
except ImportError:
    sys.exit("pip install llama-cpp-python")


def main():
    p = argparse.ArgumentParser()
    p.add_argument('--source', required=True,
                   help='Local flat-repo path or CDN URL for GGUF model')
    p.add_argument('--manifest', default='q4_0')
    p.add_argument('--cache-dir', default='./.model-cache')
    p.add_argument('--max-tokens', type=int, default=128)
    args = p.parse_args()

    prog = lambda p: print(f"\r  [{p['percent']:3d}%] {p['file'][:50]:<50s}", end='', flush=True)

    # ── 1. Resolve GGUF model ────────────────────────────────────────────
    print("=" * 60)
    print("LLM INFERENCE via model_resolver + llama-cpp-python")
    print("=" * 60)

    print(f"\nResolving GGUF model from: {args.source}")
    t0 = time.time()
    gguf_paths = resolve_gguf(args.source, manifest=args.manifest,
                              cache_dir=args.cache_dir, on_progress=prog)
    model_path = gguf_paths[0]
    print(f"\n  → {model_path}  ({time.time()-t0:.1f}s)")
    print(f"  Size: {os.path.getsize(model_path)/1048576:.1f} MB")

    # ── 2. Load model ────────────────────────────────────────────────────
    print("\nLoading model with llama-cpp-python...")
    t0 = time.time()
    llm = Llama(model_path=model_path, n_ctx=2048, n_gpu_layers=0, verbose=False)
    print(f"  ✓ Model loaded ({time.time()-t0:.1f}s)")

    # ── 3. Single-turn inference ─────────────────────────────────────────
    print("\n" + "=" * 60)
    print("SINGLE-TURN INFERENCE")
    print("-" * 60)

    prompts = [
        "What is 2 + 2? Answer in one sentence.",
        "Explain gravity to a 5 year old in two sentences.",
        "Write a haiku about programming.",
    ]

    for prompt in prompts:
        print(f"\n  User: {prompt}")
        t0 = time.time()
        result = llm.create_chat_completion(
            messages=[{"role": "user", "content": prompt}],
            max_tokens=args.max_tokens, temperature=0.7,
        )
        response = result['choices'][0]['message']['content']
        dt = time.time() - t0
        usage = result.get('usage', {})
        print(f"  Model: {response.strip()}")
        print(f"  ({dt:.1f}s, {usage.get('completion_tokens', '?')} tokens)")

    # ── 4. Multi-turn conversation ───────────────────────────────────────
    print("\n" + "=" * 60)
    print("MULTI-TURN CONVERSATION")
    print("-" * 60)

    messages = []
    turns = [
        "What is the capital of France?",
        "What is its population?",
        "And what is a famous landmark there?",
    ]

    for turn in turns:
        messages.append({"role": "user", "content": turn})
        print(f"\n  User: {turn}")
        result = llm.create_chat_completion(
            messages=messages, max_tokens=64, temperature=0.3,
        )
        reply = result['choices'][0]['message']['content'].strip()
        messages.append({"role": "assistant", "content": reply})
        print(f"  Model: {reply}")

    print("\n" + "=" * 60)
    print("✓ All LLM inference complete")
    print("=" * 60)


if __name__ == '__main__':
    main()
